core/network/src/protocol.rs

```diff
@@ -558,50 +738,75 @@ impl<B: BlockT, S: NetworkSpecialization<B>, H: ExHashT> Protocol<B, S, H> {
 			}
 		}
 
-		self.specialization.write().maintain_peers(&mut ProtocolContext::new(&self.context_data, io));
+		self.specialization
+			.maintain_peers(&mut ProtocolContext::new(
+				&mut self.context_data,
+				&self.network_chan,
+			));
 		for p in aborting {
-			io.report_peer(p, Severity::Timeout);
+			let _ = self
+				.network_chan
+				.send(NetworkMsg::ReportPeer(p, Severity::Timeout));
 		}
 	}
 
 	#[allow(dead_code)]
-	pub fn peer_info(&self, peer: NodeIndex) -> Option<PeerInfo<B>> {
-		self.context_data.peers.read().get(&peer).map(|p| {
-			PeerInfo {
-				roles: p.roles,
-				protocol_version: p.protocol_version,
-				best_hash: p.best_hash,
-				best_number: p.best_number,
-			}
+	fn peer_info(&mut self, peer: NodeIndex) -> Option<PeerInfo<B>> {
+		self.context_data.peers.get(&peer).map(|p| PeerInfo {
+			roles: p.roles,
+			protocol_version: p.protocol_version,
+			best_hash: p.best_hash,
+			best_number: p.best_number,
 		})
 	}
 
 	/// Called by peer to report status
-	fn on_status_message(&self, io: &mut SyncIo, who: NodeIndex, status: message::Status<B>) {
+	fn on_status_message(&mut self, who: NodeIndex, status: message::Status<B>) {
 		trace!(target: "sync", "New peer {} {:?}", who, status);
-
 		{
-			let mut peers = self.context_data.peers.write();
-			let mut handshaking_peers = self.handshaking_peers.write();
-			if peers.contains_key(&who) {
-				debug!(target: "sync", "Unexpected status packet from {}:{}", who, io.peer_debug_info(who));
+			if self.context_data.peers.contains_key(&who) {
+				debug!("Unexpected status packet from {}", who);
 				return;
 			}
 			if status.genesis_hash != self.genesis_hash {
-				io.report_peer(who, Severity::Bad(&format!("Peer is on different chain (our genesis: {} theirs: {})", self.genesis_hash, status.genesis_hash)));
+				let reason = &format!(
+					"Peer is on different chain (our genesis: {} theirs: {})",
+					self.genesis_hash, status.genesis_hash
+				);
+				self.network_chan.send(NetworkMsg::ReportPeer(
+					who,
+					Severity::Bad(reason.to_string()),
```

Comment:
i guess I have mentioned this already -- this file needs these `reason` extra allocations to be fixed too.